### ISSUE TITLE
Disposable decorated resources should be automatically closed on `onClose` hook.

### ISSUE THREAD

**Issue by arthurfiorette** [fastify/fastify#5669]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Since approval of [ECMAScript Explicit Resource Management](https://github.com/tc39/proposal-explicit-resource-management) and its already [adoption by Typescript 5.2](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-2.html#using-declarations-and-explicit-resource-management) and an already usage of `Symbol.asyncDispose` in https://github.com/fastify/fastify/pull/5082. We could introduce another support point of this proposal in `decorate()` where upon server close, any available `Symbol.asyncDispose` or `Symbol.dispose` would be automatically called.

### Motivation

This could improve the lifetime and resource management of resources that should be attached to an http server.

### Example

```ts
class MyDatabase {
  // Closes the database connection
  [Symbol.asyncDispose]() {}
}

app.decorateRequest('database', new MyDatabase());
```

or 

```ts
class MyRedisEventEmitter {
  // Ensures every event was sent upstream
  [Symbol.dispose]() {}
}

app.decorateRequest('emitter', new MyRedisEventEmitter());
```

or anything else that needs to be closed upon server shutdown but previously had to be done manually (and possibly forgotten).

**Comment by metcoder95** [comment]
This sounds like a good idea for a plugin that uses the `onClose` hook to dispose resources.

**Comment by climba03003** [comment]
It feels wrong to use `Symbol.asyncDispose` or `Symbol.dispose` in decorator.
The resources is actually used by the author, we have no idea the scoping of resources.
Automatically closing all the resources without knowing the scope and order is not a good idea.

Instead of decorator, providing it through plugin is more clear the scope is bind to application.

```js
function plugin() {
  fastify.decorate('database', myDatabase)
}
// treats as fastify.addHook('onClose', plugin[Symbol.dispose])
plugin[Symbol.dispose] = function() {
  this.database.close()
}

fastify.register(plugin)
```

**Comment by mcollina** [comment]
I don't think we should really attach `Symbol.dispose` to plugins, or treat those automatically in decorators. However we could support:

```
fastify.addHook('onClose', myDisposable)
```

This could easily save a few lines of code without much trouble.

**Comment by arthurfiorette** [comment]
It makes more sense! However, we would have to provide a simple wrapper for running them because async disposables do not accept parameters, and for the sync ones, we would need to manually call `done()` afterward.
